posts/onboarding.md

```diff
@@ -0,0 +1,526 @@
+            "FLOAT64": TokenType.DOUBLE,
+        }
+
+      class Parser(Parser):
+           # Specifies how certain tokens are mapped to function AST nodes
+           FUNCTIONS = {
+             **parser.Parser.FUNCTIONS,
+             "APPROX_PERCENTILE": exp.ApproxQuantile.from_arg_list,
+           }
+
+          # Specifies how a specific construct e.g. CONSTRAINT is parsed
+          def _parse_constraint(self) -> t.Optional[exp.Expression]:
+            return super()._parse_constraint() or self._parse_projection_def()
+
+    class Generator(Generator):
+        # Specifies how AST nodes, i.e. subclasses of exp.Expression, should be converted into SQL
+        TRANSFORMS = {
+            exp.Array: lambda self, e: f"[{self.expressions(e)}]",
+        }
+
+        # Specifies how AST nodes representing data types should be converted into SQL
+        TYPE_MAPPING = {
+            exp.DataType.Type.TINYINT: "INT64",
+            exp.DataType.Type.SMALLINT: "INT64",
+            ...
+        }
+```
+
+
+Even though this example is a fairly realistic starting point, we strongly encourage you to study existing dialect implementations to understand how their various components can be modified.
+
+
+## Schema
+Previous sections described the SQLGlot components used for transpilation. This section and the next describe components needed to implement other SQLGlot functionality like column-level lineage.
+
+A Schema represents the structure of a database schema, including the tables/views it contains and their column names and data types.
+
+The file `schema.py` defines the abstract classes `Schema` and `AbstractMappingSchema`; the implementing class is `MappingSchema`.
+
+A `MappingSchema` object is defined with a nested dictionary, as in this example:
+
+```Python
+schema = MappingSchema({"t": {"x": "int", "y": "datetime"}})
+```
+
+The keys of the top-level dictionary are table names (`t`), the keys of `t`’s dictionary are its column names (`x` and `y`), and the values are the column data types (`int` and `datetime`).
+
+A schema provides information required by other SQLGlot modules (most importantly, the optimizer and column level lineage).
+
+Schema information is used to enrich ASTs with actions like replacing stars (`SELECT *`) with the names of the columns being selected from the upstream tables.
+
+## Optimizer
+
+The optimizer module in SQLGlot (`optimizer.py`) is responsible for producing canonicalized and efficient SQL queries by applying a series of optimization rules.
+
+Optimizations include simplifying expressions, removing redundant operations, and rewriting queries for better performance.
+
+The optimizer operates on the abstract syntax tree (AST) returned by the parser, transforming it into a more compact form while preserving the semantics of the original query.
+
+> [!NOTE]
+> The optimizer does not currently support statistics, so it cannot employ cost-based optimizations such as join reordering
```

Comment:
It’s more than just stats, right?

sqlglot only does _logical_ optimization. 

Maybe we should make a note that the underlying engine is almost always better at optimizing for performance.

Reply:
Right, my (inverse) thought was that physical optimization is basically cost-based optimization which requires stats to do properly, but we can simply state that we only do logical optimization

Reply:
Technically, you could collect stats and write a join ordering rule in sqlglot.

So I think logical vs physical is a more precise distinction. Or in other words, sqlglot’s optimizer operates on the AST, not query plans, which is a fundamental limitation.